File: src/main/java/make/common/result/Result.java
package make.common.result;

public interface Result<T, E> {
    Object result();

    ResultType type();

    default T unwrap() throws NotOkError {
        return type().unwrap(result());
    }

    default T expected(String message) throws NotOkError {
        return type().expected(result(), message);
    }

    static <T, E> Result<T, E> err(E err) {
        return new Result<>() {
            @Override
            public Object result() {
                return err;
            }

            @Override
            public ResultType type() {
                return ResultType.ERR;
            }
        };
    }

    static <T, E> Result<T, E> ok(T ok) {
        return new Result<>() {
            @Override
            public Object result() {
                return ok;
            }

            @Override
            public ResultType type() {
                return ResultType.OK;
            }
        };
    }
}

File: src/main/kotlin/make/common/result/Result.kt
package make.common.result

sealed interface Result<out T, out E> {
    val result: Any?
    val cause: Result<*, *>?
}

data class Ok<T> @JvmOverloads
constructor(
    override val result: T,
    override val cause: Result<*, *>? = null
): Result<T, Nothing>

data class Err<E> @JvmOverloads
constructor(
    val error: E,
    override val cause: Result<*, *>? = null
): Result<Nothing, E> {
    override val result: E
        get() = error
}